### DISCUSSION TITLE
forking PortSMF

### DISCUSSION THREAD

**Discussion by Be-ing** [tenacityteam/tenacity-legacy#244] [Ideas]
PortSMF is an abandoned library from the [PortMedia project on SourceForge](https://sourceforge.net/p/portmedia/). Its README.txt says:

```
Portsmf is a relatively small number of about 9 files, so there is
currently no support for building/maintaining Portsmf as a separate
library. (Contributions are welcome.) For now, it is suggested that
you simply compile these files along with your application sources.
```
Nevertheless Debian and Arch have packages for it. Fedora does not, so in #228 I just wrote a trivial CMakeLists.txt to statically link it into Tenacity. It would be nice to give it a proper CMakeLists.txt that would make it easy to package as a separate library. I suggest we fork it as a new repository under the tenacityteam organization. This library is also used by LMMS so maybe some developers from that would be interested in collaborating on this.

**Comment by Be-ing** [comment]
ping @Veratil

**Reply by Veratil** [reply]
To mention, I don't believe it's abandoned just yet, only unable to be updated from the maintainer at this time. See the last paragraph of the README.txt, updated 2020-08-04:
```
Portmidi is long overdue for an update. Some contributions from
downstream applications have not been incorporated simply for lack of
time to do thorough cross-platform testing. I hope to fix that before
the end of the year. Please contact me if you'd like to help. -RBD
```

That being said though, it shouldn't be too hard to fork and make a library out of it. I'll help with anything if needed.

**Reply by Be-ing** [reply]
I have forked PortSMF to the tenacityteam organization: https://github.com/tenacityteam/portsmf
I have also forked PortMIDI to the mixxxdj organization and wrote a new CMakeLists.txt for it: https://github.com/mixxxdj/portmidi

**Reply by Be-ing** [reply]
I added a CMake build system to PortSMF: https://github.com/tenacityteam/portsmf/pull/1
@Veratil could you review that?

**Reply by Veratil** [reply]
> 
> 
> I added a CMake build system to PortSMF: [tenacityteam/portsmf#1](https://github.com/tenacityteam/portsmf/pull/1)
> @Veratil could you review that?

Building the library works on my test VM, so looks good there.

I'm trying out running all the tests (instead of just `test33`) using the shared library to see if anything else needs checking. So far it looks like only three tests are failing with segfaults. I'm debugging to see what's the cause or if it's something with the test itself. Other than that it seems to be working.

**Reply by Veratil** [reply]
AFAICT this is a bug in the test when creating the test sequence. It only fails in `test17`, `test19`, and `test21`. So this is something we could look into at some point, and I haven't seen this problem when reading a MIDI.

Quick detail: in `test17` specifically the duration doesn't get set, so when `seq->cut` gets called it returns `NULL` causing a segfault. I'm assuming this is similar in the other two tests (17 tests cut, 19 tests copy, 21 tests paste).

**Reply by Be-ing** [reply]
Thanks for testing! Did you modify the CMakeLists.txt to build the tests? If so please make a PR for my fork.

**Reply by Veratil** [reply]
> 
> 
> Thanks for testing! Did you modify the CMakeLists.txt to build the tests? If so please make a PR for my fork.

No, I just added some quick scripts to run the test loading the library from the build directory. Simple one-liners. I don't think it warrants adding with a commit. We'd be better off rewriting the whole testing process, and adding more.

These are run in the `portsmf_test` folder.
```sh
# make.sh
g++ -g -o portsmf_test -I../include/ -L../build -lportSMF portsmf_test.cpp

# debug.sh
LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$HOME/portsmf/build/ gdb ./portsmf_test

# run.sh
LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$HOME/portsmf/build ./portsmf_test
```

Additionally when running cmake in the `build` directory I used:
`cmake .. -DCMAKE_INSTALL_DIR=../target/ -DCMAKE_BUILD_TYPE=Debug`

**Reply by Be-ing** [reply]
I added support for building the tests and example applications, cleaned up and reorganized the repository and tagged a release. I opened a [pull request to add it to vcpkg](https://github.com/microsoft/vcpkg/pull/18982).
